src/Widgets/TerminalWidget.vala

```diff
@@ -223,6 +226,8 @@ namespace Terminal {
                 resized = true;
             });
 
+            contents_changed.connect (check_cwd_changed);
```

Comment:
The only thing that still bother me a little bit, is that this is also triggered for foreground processes. Guarding it with a `has_foreground_process` check probably doesn't make sense, because I assume they are comparatively expensive. 

Anyway it probably doesn't matter so much, I didn't see any performance differences on my machine.

Reply:
One solution would be to unregistered the callback when a foreground process is running and re-register it on `child_exited`. But this is probably a unnecessary micro optimizations, just came to my mind.

Reply:
Yes, if foreground processes cannot change the cwd then this check is unnecessary. But I doubt whether the extra complexity of code is worth the (undetectable) performance gain.